r/tests/testthat/test-expression.R

```diff
@@ -44,8 +44,13 @@ test_that("array_expression print method", {
 test_that("C++ expressions", {
   f <- FieldExpression$create("f")
   g <- FieldExpression$create("g")
+  date <- ScalarExpression$create(as.Date("2020-01-15"))
   expect_is(f == g, "ComparisonExpression")
   expect_is(f == 4, "ComparisonExpression")
+  expect_is(f == "", "ComparisonExpression")
+  expect_is(f == date, "ComparisonExpression")
+  # can't seem to make this work right now
```

Comment:
This looks like a collision between `base::Ops.Date` and our `Ops.Expression`. See "Incompatible methods" in https://stat.ethz.ch/R-manual/R-devel/library/base/html/groupGeneric.html

I'm not sure how to get around that.

Reply:
I propose resolving the other TODOs (factor, POSIXt), and we'll worry about the expression creation/Ops stuff later. I'm sitting down with Hadley next week and will review this. It would be great to have the C++ underpinnings in place so we can focus on the R evaluation.

Reply:
I'm not sure we can actually create a `DictionaryScalar` 

```c
struct ARROW_EXPORT DictionaryScalar : public Scalar {
  using Scalar::Scalar;
  using TypeClass = DictionaryType;
};
```

`Scalar` only has a constructor that holds the type, not the value.